File: sample/playground-functions/functions/src/models/push-notifications/push-notification.ts
import { classToPlain, deserialize, Expose } from "class-transformer";
import { PushNotificationType } from "./push-notification-type";

export class PushNotification {
  
  public static FromData(data: any): PushNotification {
    return deserialize(PushNotification, JSON.stringify(data));
  }

  public static FromTopic(topic: string, title: string, body: string): PushNotification {
    return new PushNotification(PushNotificationType.TOPIC, title, body, topic, Array.of());
  }

  public static FromTokens(fcmTokens: string[], title: string, body: string): PushNotification {
    return new PushNotification(PushNotificationType.TOKENS, title, body, '', fcmTokens);
  }
  
  constructor(
    type: PushNotificationType,
    title: string,
    body: string,
    topic: string,
    fcmTokens: string[]) {
    
    this.type = type;
    this.title = title;
    this.body = body;
    this.topic = topic;
    this.fcmTokens = fcmTokens;
  }
  
  public toPlain(): object {
    return classToPlain(this);
  }

  @Expose({ name: "type"})
  type: PushNotificationType;
  
  @Expose({ name: "title"})
  title: string;
  
  @Expose({ name: "body"})
  body: string;

  @Expose({ name: "topic"})
  topic: string;
  
  @Expose({ name: "fcm_tokens"})
  fcmTokens: string[];
}

File: sample/Playground/Common/Services/PushNotification/PushNotification.cs
using Microsoft.Maui.Controls.Internals;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Playground.Common.Services.PushNotification;

[Preserve(AllMembers = true)]
public sealed class PushNotification
{
    public static PushNotification FromTokens(IEnumerable<string> fcmTokens, string title, string body)
    {
        return new PushNotification(PushNotificationType.Tokens, fcmTokens: fcmTokens, title: title, body: body);
    }

    public static PushNotification FromTopic(string topic, string title, string body)
    {
        return new PushNotification(PushNotificationType.Topic, topic: topic, title: title, body: body);
    }

    private PushNotification(
        PushNotificationType type,
        IEnumerable<string> fcmTokens = null,
        string topic = null,
        string title = null,
        string body = null)
    {
        Type = type;
        FcmTokens = fcmTokens;
        Topic = topic;
        Title = title;
        Body = body;
    }

    public string ToJson()
    {
        return JsonSerializer.Serialize(this);
    }

    [JsonPropertyName("type")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public PushNotificationType Type { get; private set; }

    [JsonPropertyName("title")]
    public string Title { get; private set; }

    [JsonPropertyName("body")]
    public string Body { get; private set; }

    [JsonPropertyName("fcm_tokens")]
    public IEnumerable<string> FcmTokens { get; private set; }

    [JsonPropertyName("topic")]
    public string Topic { get; private set; }
}